### ISSUE TITLE
LBFGS with the "--output_feature_regularizer_*" option causes a segmentation fault

### ISSUE THREAD

**Issue by tzukuoh** [JohnLangford/vowpal_wabbit#818]
Running Test 16 with "--output_feature_regularizer_text" (as well as "--output_feature_regularizer_binary") causes a segmentation fault:

tkhuang@msrnyc-03:~/software/vowpal_wabbit/test$ ../vowpalwabbit/vw -k -c -d train-sets/rcv1_small.dat --loss_function=logistic -b 20 --bfgs --mem 7 --passes 20 --termination 0.001 --l2 1.0 --holdout_off --output_feature_regularizer_text r.txt
using l2 regularization = 1
enabling BFGS based optimization \*\*without\*\* curvature calculation
Num weight bits = 20
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
m = 7
Allocated 72M for weights and mem
\#\# avg. loss    der. mag.       d. m. cond.      wolfe1         wolfe2          mix fraction    curvature       dir. magnitude  step size 
creating cache_file = train-sets/rcv1_small.dat.cache
Reading datafile = train-sets/rcv1_small.dat
num sources = 1
 1 0.69315      0.00266         0.87764                                                         2.24708         776.93237       0.39057   
 3 0.51357      0.00493         0.23981          0.523903       0.088793                                        40.87817        1.00000   
 4 0.48774      0.00232         0.08215          0.274548       -0.410690                                       2.64529         1.00000   
 5 0.47879      0.00006         0.00617          0.595892       0.183063                                        0.47184         1.00000   
 6 0.47750      0.00000         0.00221          0.703360       0.403715                                        0.68626         1.00000   
 7 0.47680      0.00000         0.00038          0.588395       0.175459                                        0.08911         1.00000   
 8 0.47671      0.00000         0.00002          0.568445       0.136827  
Termination condition reached in pass 8: decrease in loss less than 0.100%.
If you want to optimize further, decrease termination threshold.
                                0.00444         1.00000   
Segmentation fault

**Comment by tzukuoh** [comment]
I created an end_examples function, which contains the only call to preconditioner_to_regularizer() (see https://github.com/tzukuoh/vowpal_wabbit/blob/master/vowpalwabbit/bfgs.cc)
But I still get a seg fault. Here is what gdb said:

**Comment by tzukuoh** [comment]
I created an 'end_examples' function, which contains the only call to preconditioner_to_regularizer(). (see https://github.com/tzukuoh/vowpal_wabbit/blob/master/vowpalwabbit/bfgs.cc). But still got a seg fault. According to gdb (see log below), the seg fault occurs in save_load_regularizer. Looks like end_examples() has not been called at that point? 

(gdb) run -k -c -d train-sets/rcv1_small.dat --loss_function=logistic -b 20 --bfgs --mem 7 --passes 20 --termination 0.001 --l2 1.0 --holdout_off --output_feature_regularizer_text r.txt
Starting program: /home/NORTHAMERICA/tkhuang/software/vowpal_wabbit/vowpalwabbit/vw -k -c -d train-sets/rcv1_small.dat --loss_function=logistic -b 20 --bfgs --mem 7 --passes 20 --termination 0.001 --l2 1.0 --holdout_off --output_feature_regularizer_text r.txt
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
using l2 regularization = 1
enabling BFGS based optimization **without** curvature calculation
Num weight bits = 20
learning rate = 0.5
initial_t = 0
power_t = 0.5
decay_learning_rate = 1
m = 7
Allocated 72M for weights and mem
## avg. loss    der. mag.       d. m. cond.      wolfe1         wolfe2          mix fraction    curvature       dir. magnitude  step size
creating cache_file = train-sets/rcv1_small.dat.cache
Reading datafile = train-sets/rcv1_small.dat
num sources = 1 
[New Thread 0x7ffff2026700 (LWP 18533)]          
 1 0.69315      0.00266         0.87764                                                         2.24708         776.93237       0.39057
 3 0.51357      0.00493         0.23981          0.523903       0.088793                                        40.87817        1.00000
 4 0.48774      0.00232         0.08215          0.274548       -0.410690                                       2.64529         1.00000
 5 0.47879      0.00006         0.00617          0.595892       0.183063                                        0.47184         1.00000
 6 0.47750      0.00000         0.00221          0.703360       0.403715                                        0.68626         1.00000
 7 0.47680      0.00000         0.00038          0.588395       0.175459                                        0.08911         1.00000
 8 0.47671      0.00000         0.00002          0.568445       0.136827
Termination condition reached in pass 8: decrease in loss less than 0.100%.
If you want to optimize further, decrease termination threshold.
                                0.00444         1.00000

Program received signal SIGSEGV, Segmentation fault.
0x00000000004b801e in save_load_regularizer (all=..., b=..., model_file=..., read=false, text=true) at bfgs.cc:880
880               if (*v != 0.)
(gdb) bt
#0  0x00000000004b801e in save_load_regularizer (all=..., b=..., model_file=..., read=false, text=true) at bfgs.cc:880
#1  0x00000000004b865e in save_load (b=..., model_file=..., read=false, text=true) at bfgs.cc:955
#2  0x0000000000427f3a in LEARNER::learner<char>::save_load (this=0x820570, io=..., read=false, text=true) at learner.h:136
#3  0x0000000000425494 in dump_regressor (all=..., reg_name=..., as_text=true) at parse_regressor.cc:340
#4  0x00000000004257bd in finalize_regressor (all=..., reg_name=...) at parse_regressor.cc:365
#5  0x00000000004b7cdc in end_pass (b=...) at bfgs.cc:802
#6  0x0000000000493c42 in LEARNER::learner<char>::end_pass (this=0x820570) at learner.h:159
#7  0x0000000000493874 in LEARNER::generic_driver (all=...) at learner.cc:29
#8  0x0000000000421616 in main (argc=20, argv=0x7fffffffe518) at main.cc:70

**Comment by JohnLangford** [comment]
I believe this is fixed now.
